### DISCUSSION TITLE
Will GrandOrgue work headless?

### DISCUSSION THREAD

**Discussion by William-Purvis** [GrandOrgue/grandorgue#1499] [Help]
I'm developing an organ with a normal setup, but when it's finished, I want it to operate without monitor, PC keyboard or mouse.
I've got it set up so that it comes up OK from power-on, but so far only with display attached. Does anyone know what happens
if no display is detected at boot up? I guess I'm hoping it will just start up regardless. The display could be simply unplugged, and 
if maintenance or updates are needed I can plug the display, keyboard and mouse in before powering on.

**Comment by William-Purvis** [comment]
As no-one has come up with an immediate answer, I tried it out myself. Powered off the display, then booted up the PC.
After a minute or so to allow booting to complete, I logged in using SSH from my laptop, and GrandOrgue appeared to be running.
Pressing keys on the manual produce sounds so the answer seems to be YES.

**Reply by Aduragbemi-Adeyinka** [reply]
Yes GrandOrgue can run headless. I have a setup where the CPU is sealed into the console. I extended the power-up ternimal to the knee-board, connected to a push button. Just press the button briefly; GrandOrgue and other apps that I need to run the VPO launch on power-up, and the last used organ loads automatically. The largest organ that I use (Friesach) typically loads in less than 1 minute. The keypad and mouse are wireless so I put them away. So is the display monitor, except when I need to make a change in settings.

![Screenshot_20230427-090055_Chrome](https://github.com/GrandOrgue/grandorgue/assets/105753578/ce6ac6ff-d591-4d8f-8012-3a9f32eea247)

**Reply by bascouat** [reply]
Hello, great to find that information. What pc do you use, I would like to get the cheapest stable one for use of Friesach 3 manuals only, that should run as you two have. Thank you!

**Reply by newVideophile** [reply]
I come a month later, but yes, GO can work headless. I use a Raspberry Pi 4 with 8 Gb RAM, which probably used to be the cheapest possible PC that can run Friesach. I say "used to", because the Pi has been out of stock and the price of the few available units has become stellar...

Running GO headless has some limitations, though: some features cannot be controlled through MIDI and thus it is not possible to access them in headless mode. In fact, GO has not been designed to be used headless. This is unfortunate, because using a keyboard, a mouse or a touchpanel while playing is not convenient.

**Reply by rousseldenis** [reply]
@newVideophile I opened this conversation. Maybe it will be in your interest https://github.com/GrandOrgue/grandorgue/issues/1563

**Comment by William-Purvis** [comment]
On 06/06/2023 12:48, newVideophile wrote:
>
> I come a month later, but yes, GO can work headless. I use a Raspberry 
> Pi 4 with 8 Gb RAM, which probably used to be the cheapest possible PC 
> that can run Friesach. I say "used to", because the Pi has been out of 
> stock and the price of the few available units have become stellar...
>
> Running GO headless has some limitations, though: some features cannot 
> be controlled through MIDI and thus it is not possible to access them 
> in headless mode. In fact, GO has not been designed to be used 
> headless. This is unfortunate, because using a keyboard, a mouse or a 
> touchpanel while playing is not convenient.
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/GrandOrgue/grandorgue/discussions/1499#discussioncomment-6100106>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AH27WMZ6FSIPZBPNNEK6P6LXJ4KJZANCNFSM6AAAAAAXR4O5TE>.
> You are receiving this because you authored the thread.Message ID: 
> ***@***.***>
>
I've already established that it works fine in headless mode. Ran a few 
tests which worked OK.
It's intended to be used as a church organ, with one ODF, and so 
'normal' use is just switch on,
play, then switch off. GO does all that fine. I do have the option of 
plugging in monitor, keyboard and mouse
to do system updates (hopefully none as it runs without any networking) 
and possible changes to the ODF,
addition of other sample files, etc. I may add the ability to play from 
a USB Memory stick MIDI file,
but that's not in the requirement.

It's looking good so far! Just got to complete building the console. 
Spent the morning varnishing....

Bill

-- 
+----------------------------------------+
| Bill Purvis                            |
| ***@***.***                   |
+----------------------------------------+

**Comment by powerinpete** [comment]
I also use GO on a headless church console. I'm wondering if it's possible to have a panel in GO (like the coupler or master control panels for example) with controls/buttons that have user-definable functions. A simple example might be to be able to assign a keyboard shortcut to a panel control...so assigning Ctrl+S to a panel button would enable you to Save organ settings using a physical control on the console that has been mapped to that panel button.

A couple of things that would be useful to me would be being able to change the transposer and having a few buttons that could be pre-programmed to load different sample sets (perhaps from the Favorites list)...using a piston or other midi switch mapped on the physical console.

**Reply by kerkovits** [reply]
These are all already implemented:

Transposer and save is available as MIDI pistons at Panel→Master controls.
Assigning MIDI messages to load an organ is available at File→Settings→Organs, choose the organ to load, and click on button MIDI.

**Reply by powerinpete** [reply]
Thanks for that!

I see now that I could have found all this by re-reading the whole help file (haven't read it all for several years)...but when trying to find something specific sometimes it's a matter of knowing what to look for or knowing what search terms to use. I was fooled a bit by testing what I thought were (and are) the master control transpose buttons in GO on my home computer but they didn't work...perhaps because I have no MIDI keyboards connected to it. I also wrongly assumed the Save button was just to save the tuning settings :-P

Trying now on my home organ (which is also usually headless) they work perfectly. Thanks again!
